Expensify/App
# Add accountID to the set password link

### Details
The `SetPassword` API command requires an email to be provided, along with a new password and a validateCode. For the `SetPasswordPage`, we were previously assuming that the email would be read from Onyx storage. 

However, that may not be the case if we send the user a set password link via a notification, as the user may not have completed the first part of the sign-in flow (which saves their email into Onyx storage). If this is the case, the API call to `SetPassword` will fail since the email sent over will be `undefined`.

In lieu of the changes in https://github.com/Expensify/Web-Expensify/pull/30466/ and https://github.com/Expensify/Auth/pull/5402, I'm adding an email parameter to the set password route to ensure we're always sending over and reading an email when the user navigates from this page.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/156786
$ https://github.com/Expensify/Expensify/issues/157873

### Tests (local dev)
1. On Web-Expensify, check out `jasper-ecashValidatePasswordEmail`; on Auth, check out and make `jasper-resendValidateCodeSetCashPassword`.
2. Make sure you're following the steps with an unvalidated Expensify email. Run `./script/sql.sh` then:
- Get your accountID:
```
SELECT validated from accounts WHERE email = "[your email]@expensify.com";
```
- Make sure your account is invalidated with this command:
```
UPDATE logins SET validatedDate = "" WHERE accountID = [your accountID];
``` 
- Make sure you've got the most up to date partnerName for Expensify.cash:
```
INSERT OR IGNORE INTO partners ( created, partnerID, partnerName, partnerPassword, partnerKey ) VALUES ( '2021-03-18 00:00:00', 84, 'chat-expensify-com', 'DD1431EB9E11F0071EC05C24E9CB2FD595B1ED75', 'asdf,fdsa');
```
3. Hit `expensify.com.dev/api?command=ResendValidateCode&referer=https://expensify.cash/&accountID=[your accountID]` to queue up a job that will send yourself a set password email.
- Use `SELECT * from notifications;` and check to make sure there's a `SetCashPassword` notification with your accountID queued up in the last row.
4. Then run `vssh` and `php /vagrant/Web-Expensify/script/notifyall.php` to send it out.
- to troubleshoot bedrock jobs, use the [dev Bedrock Job Manager](https://www.expensify.com.dev/_devportal/tools/bedrockJobManager/)
5. Check your email for an email with the subject "Your magic sign in link for Expensify.cash". Click on the link.
6. Modify the first part of the URL with whatever localhost:[port] your local version of cash is running on (e.g. localhost:8081)
- After this, the URL should look something like this: localhost:8081/setpassword/[accountID]/[validateCode]
7. Navigate to that new URL and set a new password (it's better to do this in an incognito window).
8. Sign out of e.cash and try to sign in to that email with the new password. Verify that you can sign in.

<img width="1010" alt="Screen Shot 2021-04-07 at 9 44 39 AM" src="https://user-images.githubusercontent.com/31285285/113798513-e26a6480-9785-11eb-80ac-3acb224188dc.png">

### QA Test Steps
1. Navigate to the sign in page.
2. Create a new account with an email address you have access to (I recommend using [EmailOnDeck](https://www.emailondeck.com))
3. Verify that an email (like the screenshot) was sent to that email address.
4. Click "Resend Link". Verify that another email was sent to that email address.
5. In the newer email, click `here`. Verify that it takes you to `https://expensify.cash/setpassword/:accountID/:validateCode`
6. Set a password, it should log you in after you click "Set Password"
7. Sign out. Verify that you can sign back in with the same email address and password you just set.
- NOTE: this can also be tested on iOS/Android

### Tested On


### Screenshots
#### Web
<img width="1792" alt="Screen Shot 2021-03-18 at 11 19 24 AM" src="https://user-images.githubusercontent.com/31285285/111568490-cfdeba00-87db-11eb-81da-7783b36f7cc1.png">

#### Mobile Web
<img width="462" alt="Screen Shot 2021-03-18 at 11 18 40 AM" src="https://user-images.githubusercontent.com/31285285/111568496-d3724100-87db-11eb-9b72-b19b3a87afae.png">
